### PR TITLE
feat(dataframe): support constructing DataFrame from list

### PR DESCRIPTION
**Pull request by AChenQ**

### CODE REVIEW

**Comment by GravitiRob** [comment]
Task linked: [CU-28v2czn 基于 PagingList 实现 DataFrame.__init__ 方法](https://app.clickup.com/t/28v2czn)

File: graviti/dataframe/frame.py [hunk 1]
```diff
@@ -67,31 +68,36 @@ class DataFrame:
...
-        if isinstance(data, dict):
-            for key, value in data.items():
-                self._columns[key] = (
-                    DataFrame(value) if isinstance(value, dict) else ColumnSeries(value, name=key)
-                )
-                self._column_names.append(key)
-            self._index = ColumnSeries(list(range(self.__len__())))
-        else:
-            raise ValueError("DataFrame only supports generating from dictionary now")
+        for key in array[0].keys():
```
Comment (linjiX) [inline graviti/dataframe/frame.py:92]: why we need to get the first element of array?
what if the array is an empty array?

File: graviti/dataframe/frame.py [hunk 2]
```diff
@@ -67,31 +68,36 @@ class DataFrame:
...
         if columns is not None:
             # TODO: missing columns processing
             pass
 
+        if not isinstance(data, DataFrame) and isinstance(data, Iterable):
+            self._init(pa.array(data))
+        else:
+            raise ValueError("DataFrame only supports generating from Sequence now")
+
+    def _init(self, array: pa.StructArray) -> None:
```
Comment (linjiX) [inline graviti/dataframe/frame.py:89]: is possible to use `__new__` to remove `_init`?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/column/series.py [hunk 3]
```diff
@@ -40,16 +44,19 @@ class Series:
...
         index: Optional[Iterable[int]] = None,
     ) -> None:
-        if schema is not None:
-            # TODO: missing schema processing
-            pass
+        if schema is None:
+            raise NotImplementedError("Inferred schema from `data` is not support now.")
+
+        if not isinstance(data, pa.Array):
+            data = pa.array(data)
```
Comment (linjiX) [inline graviti/dataframe/column/series.py:56]: Shall we check the data fit the schema

File: graviti/dataframe/frame.py [hunk 4]
```diff
@@ -168,7 +175,39 @@ def from_lazy_lists(cls: Type[_T], lazy_lists: "LazyLists") -> _T:
             The loaded :class:`~graviti.dataframe.DataFrame` object.
 
         """
-        return cls(lazy_lists)
+        # Not support loading tensorbay Dataset.
+        return cls(lazy_lists)  # type: ignore[arg-type]
+
+    @classmethod
+    def from_pyarrow(cls: "Type[DataFrame]", array: pa.StructArray, schema: record) -> "DataFrame":
```
Comment (linjiX) [inline graviti/dataframe/frame.py:182]: we should use TypeVar for cls and return type,
if DataFrame has a subclass and we use "DataFrame" as the Type, the type will mismatch in subclass

File: graviti/dataframe/frame.py [hunk 5]
```diff
@@ -168,7 +175,39 @@ def from_lazy_lists(cls: Type[_T], lazy_lists: "LazyLists") -> _T:
...
+            The loaded :class:`~graviti.dataframe.DataFrame` object.
+
+        """
+        obj: "DataFrame" = object.__new__(cls)
+        # TODO: Check the schema is valid for pyarrow struct array.
+        obj.schema = schema
+        obj._columns = {}
+        obj._column_names = []
+
+        for key in array.type:
```
Comment (linjiX) [inline graviti/dataframe/frame.py:199]: `key` - > `field` or `pafield`?

File: graviti/dataframe/frame.py [hunk 6]
```diff
@@ -60,38 +57,45 @@ class DataFrame:
...
+                obj._columns = {}
+                obj._column_names = []
+                obj._index = ColumnSeries([])
+                obj.schema = record([])
+                return obj
+
+            raise NotImplementedError("Inferred schema from `data` is not support now.")
+
+        if not isinstance(data, DataFrame) and isinstance(data, Iterable):
+            return DataFrame.from_pyarrow(pa.array(data), schema)
```
Comment (linjiX) [inline graviti/dataframe/frame.py:96]: you should use the schema type to generate the array.
`pa.array(data, schema.to_pyarrow())`

File: graviti/dataframe/frame.py [hunk 7]
```diff
@@ -168,7 +175,39 @@ def from_lazy_lists(cls: Type[_T], lazy_lists: "LazyLists") -> _T:
...
+        obj._columns = {}
+        obj._column_names = []
+
+        for key in array.type:
+            column_name = key.name
+            column_value = array.field(column_name)
+            column_schema = schema.fields[column_name]
+            obj._columns[column_name] = (
+                DataFrame.from_pyarrow(column_value, schema=column_schema)
+                if isinstance(column_value, pa.StructArray) and isinstance(column_schema, record)
```
Comment (linjiX) [inline graviti/dataframe/frame.py:205]: shall we check the schema type rather than array type?

File: graviti/dataframe/frame.py [hunk 8]
```diff
@@ -168,7 +175,39 @@ def from_lazy_lists(cls: Type[_T], lazy_lists: "LazyLists") -> _T:
...
+                else ColumnSeries(column_value, name=column_name, schema=column_schema)
+            )
+            obj._column_names.append(column_name)
+        obj._index = ColumnSeries(list(range(obj.__len__())))
```
Comment (linjiX) [inline graviti/dataframe/frame.py:209]: will `len(column_value)` better than `obj.__len__()`

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/frame.py [hunk 9]
```diff
@@ -168,7 +176,38 @@ def from_lazy_lists(cls: Type[_T], lazy_lists: "LazyLists") -> _T:
...
+        obj._columns = {}
+        obj._column_names = []
+
+        for pafield in array.type:
+            column_name = pafield.name
+            column_value = array.field(column_name)
+            column_schema = schema.fields[column_name]
+            obj._columns[column_name] = (
+                DataFrame.from_pyarrow(column_value, schema=column_schema)
+                if isinstance(pafield.type, pa.StructType) and isinstance(column_schema, record)
```
Comment (linjiX) [inline graviti/dataframe/frame.py:205]: why we need to check `column_schema` here?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/frame.py [hunk 10]
```diff
@@ -168,7 +176,38 @@ def from_lazy_lists(cls: Type[_T], lazy_lists: "LazyLists") -> _T:
...
+                else ColumnSeries(column_value, name=column_name, schema=column_schema)
+            )
+            obj._column_names.append(column_name)
+        obj._index = ColumnSeries(list(range(len(obj))), schema=int32())
```
Comment (linjiX) [inline graviti/dataframe/frame.py:209]: should ColumnSeries accept `Iterable`? do we need to transfer range to list?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/frame.py (hunk 9 above)
Comment (AChenQ) [reply graviti/dataframe/frame.py:205]: for mypy checking.

**Review by linjiX — CHANGES_REQUESTED** [review]

**Comment by linjiX** [comment]
Implemented in https://github.com/Graviti-AI/graviti-python-sdk/pull/201

**Pull request closed without merging**